### ISSUE TITLE
How do I attach VS Code to a dockerd container running inside WSL2 - Ubuntu? [Question]

### ISSUE THREAD

**Issue by amit-agni** [microsoft/vscode-remote-release#3254]
My VS Code version is 1.45.1 (user setup). It is running on Win 10 Build 20150.1000

On Win 10 I have WSL2 enabled. On WSL2 I have a Ubuntu 18.04 distribution and within Ubuntu I have docker daemon running a container.  

How do I attach VS Code to this container ?

I can use the Remote-WSL extension to connect to Ubuntu but how do I attach VS Code to the container that is running within Ubuntu ?

NOTE : I could have used Docker Desktop on Win 10 but I am using the [CUDA development release on WSL2](https://docs.nvidia.com/cuda/wsl-user-guide/index.html#setting-containers), so I have to use the dockerd in Ubuntu and cannot use Docker desktop. This is related to [Microsoft's support for CUDA in WSL2](https://docs.microsoft.com/en-au/windows/win32/direct3d12/gpu-cuda-in-wsl)

_Please excuse me if this is not the right forum for this question._

**System Info :**
VS Code version is 1.45.1 (user setup)
Win 10 Pro Build 20150.1000
WSL2 :
Kernel Version: 4.19.121-microsoft-standard
Operating System: Ubuntu 18.04.4 LTS

**Comment by PavelSosin-320** [comment]
Technically it is impossible because Attach means "attach to the TTy in/out/err" but VSCode and your daemon run in different VMs. If your Docker keeps port 2375 open and accessible via TCP w/o TLS from Windows PS or CMD, you can attach to your container using docker -H .... Please, take into account that your WSL Distro IP is different from your machine IP so you need some Windows defender arrangement. 
By default, Docker Desktop daemon listens pipe protocol from the Windows side and Unix socket protocol from the Ubuntu distro side.
What you can do is 1. Remote terminal to your distro and attach to your container using Docker default context of your distro.

**Comment by amit-agni** [comment]
Thanks for the detailed explanation. 

This is what I tried, with some help from [here](https://docs.docker.com/engine/reference/commandline/dockerd/). 
On WSL/Ubuntu, opened the port 2375 as you suggested :
`sudo dockerd -H unix:///var/run/docker.sock -H tcp://0.0.0.0:2375`

And then on Win10 PS/CMD (without starting the Docker destop) 
Started the container : `docker -H tcp://172.29.145.126:2375 start tf_gpu1`
And then attached the container :  `docker -H tcp://172.29.145.126:2375 attach tf_gpu1`
_where 172.29.145.126 is the ip address of WSL/Ubuntu_


Output on WSL/Ubuntu :
`CONTAINER ID        IMAGE                                  COMMAND                  CREATED             STATUS                    PORTS                    NAMES`
`c86e253552b6        tf121_gpu                              "jupyter notebook --…"   4 minutes ago       Up 38 seconds             8888/tcp                 tf_gpu1`

Output on Win 10 :
`PS C:\Users\amit> docker ps -a`
`error during connect: Get http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.40/containers/json?all=1: open //./pipe/docker_engine: The system cannot find the file specified. In the default daemon configuration on Windows, the docker client must be run elevated to connect. This error may also indicate that the docker daemon is not running.`

I am not sure what is expected. I thought when I ran `docker ps -a` on Win 10 PS, I would get to see the container that is created in WSL\Ubuntu. And then I could use VS Code to attach to that "running" container ?

Maybe I missed this "attach to your container using Docker default context of your distro." How to do it correctly ?
